design/host-discovery-controller/host-discovery.md

```diff
@@ -0,0 +1,325 @@
+`baremetal-operator` to only look for hosts with the flag set. It
+isn’t really clear how useful that would be, so the description above
+does not assume such a limit.
+
+Nodes are removed from Ironic when the host CR is deleted. A host may
+be rediscovered if the data is deleted and the host is rebooted into
+the Ironic agent image. This could cause confusion, if the users
+intent was to remove the host completely. It could also be seen as a
+feature, if it is considered a way to reprovision a host.
+
+### Work Items
+
+1. Create the new git repository for the controller and API and set up
+   CI jobs.
+2. Write the new controller
+3. Update ironic-inspector-image to add unknown hosts using the `fake`
+   driver instead of the `ipmi` driver so they can be provisioned
+   without BMC credentials.
+4. Update `IronicProvisioner` to allow it to update an existing node
+   in Ironic so that discovered nodes will have their settings
+   adjusted, as needed.
+5. Update the state machine to add `Importing` state, and make the
+   related changes to the `IronicProvisioner` API to allow it to look
+   for inspection data without triggering the inspection process.
+6. Update the state machine to support deleting hosts in the
+   `Importing` state, including updating the enum for the host CRD to
+   allow the new state.
+7. Update `IronicProvisioner` so it will recognize a manually
+   registered host that matches a discovered node in the Ironic
+   database.
+8. Document how to fully enable and disable the feature, including the
+   fact that hosts need to be configured to boot from disk and only
+   fall back to PXE if there is no image on the drive.
+9. Extend the `DemoProvisioner` to support the new states.
+10. Add a method to `BareMetalHost` to report whether power control
+   operations are supported. For now, this will only look at whether
+   there are BMC credentials.
+11. Update the bare metal controller to skip trying to apply power
+   state management operations to hosts that do not support them.
+
+### Dependencies
+
+[Ironic node discovery spec](https://specs.openstack.org/openstack/ironic-inspector-specs/specs/ironic-node-auto-discovery.html)
+
+### Test Plan
+
+Add unit tests for the new controller.
+
+Add unit tests for the state transitions using the test provisioner.
+
+Extend the integration test suite?
+
+### Upgrade / Downgrade Strategy
+
+Updating the host CRD to support the new state is a breaking
+change. Downgrading a system with a host in the `Importing` state may
+trigger failures in the `baremetal-operator` because it will not
+recognize the new host state.
+
+### Version Skew Strategy
+
+N/A
+
+## Drawbacks
+
+N/A
+
+## Alternatives
+
+Instead of polling the Ironic API looking for new hosts, we could
+watch the `dnsmasq` logs. However, in some configurations PXE may be
+disabled or hosts may be booted in other ways even with PXE enabled,
+so we would not see all discovered hosts.
```

Comment:
Strictly speaking DHCP does not imply PXE, but somebody could be using their own DHCP + ISO boot.